Added htmlParser.element.findOne

engineering-this opened the pull request:
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [ ] Manual tests

## What changes did you make?

- Added `CKEDITOR.htmlParser.element.findOne`
- Refactored Paste From Word filter function `removeSymbolText`.

Closes #2698.

Comandeer commented on tests/core/htmlparser/element.js:
```diff
@@ -16,6 +16,14 @@ function write( el ) {
 
+var testHtml = '<body>' +
+		'<p>Foo' +
+			'<div>Baz' +
```
This HTML is incorrect: `div` can't be nested in `p` tag.

engineering-this commented on core/htmlparser/element.js:
```diff
@@ -470,6 +470,38 @@ CKEDITOR.htmlParser.cssStyle = function() {
+		 */
+		findOne: function( criteria ) {
+			var findMethod = typeof criteria === 'function' ? criteria : findByName,
+				nestedMatch,
```
Yes, but I'm using it inside next declaration, so moving it to the end would break linter rule which requires variable declaration before it is used.

engineering-this commented on core/htmlparser/element.js:
```diff
@@ -470,6 +470,38 @@ CKEDITOR.htmlParser.cssStyle = function() {
+				match = CKEDITOR.tools.array.find( this.children, function( child ) {
+					if ( findMethod( child ) ) {
+						return true;
+					} else if ( child.children && child.findOne ) {
```
We need `else if` branch, because child can be `CKEDITOR.htmlParser.text` which doesn't have property `children` or method `findOne`.

engineering-this commented:
- Changed returned type to `null` when no match is found.
- Extracted `findOne` tests to separate file, so they are much better readable with html string right above TCs. Also reformatted fixture, so each node is in new line, except for elements that contain only one text.
- Corrected changelog, and other small things.

Comandeer commented on plugins/pastefromword/filter/default.js:
```diff
@@ -1172,25 +1172,21 @@
-					}
+				node = element.findOne( function( node ) {
+					return node.value && node.value.indexOf( symbol ) > -1;
+				}, true ),
```
Code style issue – there should be one more indentation level.

engineering-this replied:
Adding indentation level would make closing bracket higher indentation than opening, shouldn't they be the same?

Comandeer replied:
It's a very similar case to objects in such lists and they are indented with that extra tab: https://github.com/ckeditor/ckeditor-dev/blob/1f365357c043a74e69ad21f32c829929d8610106/plugins/tableselection/plugin.js#L893-L901

Comandeer commented on plugins/pastefromword/filter/default.js:
```diff
@@ -1172,25 +1172,21 @@
-
-			element.forEach( function( node ) {
-				// Since symbol may contains special characters we use `indexOf` (instead of RegExp) which is sufficient (#877).
-				// Stop after after first replaced string (#2690).
```
Why these comments are removed? They still seem to be accurate.

Comandeer commented on tests/core/htmlparser/elementfindone.js:
```diff
@@ -0,0 +1,44 @@
+( function() {
+	'use strict';
+
+	var testElement = CKEDITOR.htmlParser.fragment.fromHtml(
```
I wonder if it wouldn't be more readable to move the HTML into separate file, `elementfindone.html`.

Comandeer commented on tests/core/htmlparser/elementfindone.js:
```diff
@@ -0,0 +1,44 @@
+					return element.attributes && element.attributes[ 'class' ] === 'faz';
+				} )
+			);
+			assert.areSame( testElement.children[ 0 ].children[ 1 ].children[ 1 ],
```
In such tests it'd be nice to add some fail message to know, which elements are not equal.

Comandeer commented on plugins/pastefromword/filter/default.js:
```diff
@@ -1172,25 +1172,23 @@
 
-			element.forEach( function( node ) {
+			if ( node ) {
 				// Since symbol may contains special characters we use `indexOf` (instead of RegExp) which is sufficient (#877).
```
`indexOf` was moved up, so the comment also should.

Comandeer commented on core/htmlparser/element.js:
```diff
@@ -470,6 +470,42 @@ CKEDITOR.htmlParser.cssStyle = function() {
+		 * @param {String/Function} criteria Tag name or evaluator function.
+		 * @param {CKEDITOR.htmlParser.node} criteria.child The currently iterated child.
+		 * @param {Boolean} [recursive=false] If set to `true` will iterate all descendants, otherwise only direct children.
+		 * @returns {CKEDITOR.htmlParser.node} First matched child, `null` otherwise.
```
It should be marked as nullable type.

Comandeer commented on core/htmlparser/element.js:
```diff
@@ -470,6 +470,42 @@ CKEDITOR.htmlParser.cssStyle = function() {
+		 * @returns {CKEDITOR.htmlParser.node} First matched child, `null` otherwise.
+		 */
+		findOne: function( criteria, recursive ) {
+			var findMethod = typeof criteria === 'function' ? criteria : function( item ) {
```
By inlining I meant "inline it directly into `CKEDITOR.tools.array.find` callback":

```javascript
var isMatching = typeof criteria === 'function' ? criteria( child ) : item.name === criteria;
```

IMO such version is still readable and dedicated function `findMethod` is not necessary in this case.

engineering-this commented:
Fixed failing test. Fail message was wrongly placed as `findOne` param instead of `assert`, as a result `findOne` treated message as `recursive` param, and `assert` didn't have any message to display.

engineering-this commented:
IE cuts some of text nodes. To make test constant regardless of environment I've moved tested html code back into js file and removed html file.

engineering-this commented:
These tests fails on major too -> #2806
